File: nimo-plm/internal/model/entity/bom.go
package entity

import (
	"time"
)

// BOMHeader BOM头表
type BOMHeader struct {
	ID           string     `json:"id" gorm:"primaryKey;size:32"`
	ProductID    string     `json:"product_id" gorm:"size:32;not null"`
	Version      string     `json:"version" gorm:"size:16;not null"`
	Status       string     `json:"status" gorm:"size:16;not null;default:draft"`
	Description  string     `json:"description" gorm:"type:text"`
	TotalItems   int        `json:"total_items" gorm:"not null;default:0"`
	TotalCost    float64    `json:"total_cost" gorm:"type:decimal(15,4)"`
	MaxLevel     int        `json:"max_level" gorm:"not null;default:0"`
	ReleasedBy   string     `json:"released_by" gorm:"size:32"`
	ReleasedAt   *time.Time `json:"released_at"`
	ReleaseNotes string     `json:"release_notes" gorm:"type:text"`
	CreatedBy    string     `json:"created_by" gorm:"size:32;not null"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`

	// 关联
	Product  *Product   `json:"product,omitempty" gorm:"foreignKey:ProductID"`
	Creator  *User      `json:"creator,omitempty" gorm:"foreignKey:CreatedBy"`
	Releaser *User      `json:"releaser,omitempty" gorm:"foreignKey:ReleasedBy"`
	Items    []BOMItem  `json:"items,omitempty" gorm:"foreignKey:BOMHeaderID"`
}

func (BOMHeader) TableName() string {
	return "bom_headers"
}

// BOMItem BOM行项
type BOMItem struct {
	ID           string    `json:"id" gorm:"primaryKey;size:32"`
	BOMHeaderID  string    `json:"bom_header_id" gorm:"size:32;not null"`
	ParentItemID string    `json:"parent_item_id" gorm:"size:32"`
	MaterialID   string    `json:"material_id" gorm:"size:32;not null"`
	Level        int       `json:"level" gorm:"not null;default:0"`
	Sequence     int       `json:"sequence" gorm:"not null;default:0"`
	Quantity     float64   `json:"quantity" gorm:"type:decimal(15,4);not null"`
	Unit         string    `json:"unit" gorm:"size:16;not null;default:pcs"`
	Position     string    `json:"position" gorm:"size:32"`
	Reference    string    `json:"reference" gorm:"size:128"`
	Notes        string    `json:"notes" gorm:"type:text"`
	UnitCost     float64   `json:"unit_cost" gorm:"type:decimal(15,4)"`
	ExtendedCost float64   `json:"extended_cost" gorm:"type:decimal(15,4)"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`

	// 关联
	BOMHeader  *BOMHeader `json:"bom_header,omitempty" gorm:"foreignKey:BOMHeaderID"`
	ParentItem *BOMItem   `json:"parent_item,omitempty" gorm:"foreignKey:ParentItemID"`
	Material   *Material  `json:"material,omitempty" gorm:"foreignKey:MaterialID"`
	Children   []BOMItem  `json:"children,omitempty" gorm:"foreignKey:ParentItemID"`
}

func (BOMItem) TableName() string {
	return "bom_items"
}

// BOMStatus BOM状态
const (
	BOMStatusDraft    = "draft"
	BOMStatusReleased = "released"
	BOMStatusObsolete = "obsolete"
)

File: nimo-plm-web/src/api/bom.ts
import apiClient from './client';
import { ApiResponse } from '@/types';
import { Material } from './materials';

export interface BOMItem {
  id: string;
  bom_header_id: string;
  parent_item_id: string;
  material_id: string;
  level: number;
  sequence: number;
  quantity: number;
  unit: string;
  position: string;
  reference: string;
  notes: string;
  unit_cost: number;
  extended_cost: number;
  created_at: string;
  updated_at: string;
  material?: Material;
}

export interface BOMHeader {
  id: string;
  product_id: string;
  version: string;
  status: string;
  description: string;
  total_items: number;
  total_cost: number;
  max_level: number;
  released_by: string;
  released_at: string;
  release_notes: string;
  created_by: string;
  created_at: string;
  updated_at: string;
  items?: BOMItem[];
}

export interface BOMVersionsResponse {
  versions: BOMHeader[];
}

export const bomApi = {
  get: async (productId: string): Promise<BOMHeader> => {
    const response = await apiClient.get<ApiResponse<BOMHeader>>(`/products/${productId}/bom`);
    return response.data.data;
  },

  listVersions: async (productId: string): Promise<BOMVersionsResponse> => {
    const response = await apiClient.get<ApiResponse<BOMVersionsResponse>>(`/products/${productId}/bom/versions`);
    return response.data.data;
  },

  addItem: async (productId: string, data: Partial<BOMItem>): Promise<BOMItem> => {
    const response = await apiClient.post<ApiResponse<BOMItem>>(`/products/${productId}/bom/items`, data);
    return response.data.data;
  },

  updateItem: async (productId: string, itemId: string, data: Partial<BOMItem>): Promise<BOMItem> => {
    const response = await apiClient.put<ApiResponse<BOMItem>>(`/products/${productId}/bom/items/${itemId}`, data);
    return response.data.data;
  },

  deleteItem: async (productId: string, itemId: string): Promise<void> => {
    await apiClient.delete(`/products/${productId}/bom/items/${itemId}`);
  },

  release: async (productId: string, data: { version: string; release_notes?: string }): Promise<BOMHeader> => {
    const response = await apiClient.post<ApiResponse<BOMHeader>>(`/products/${productId}/bom/release`, data);
    return response.data.data;
  },

  compare: async (productId: string, params: { version1: string; version2: string }): Promise<any> => {
    const response = await apiClient.get(`/products/${productId}/bom/compare?version1=${params.version1}&version2=${params.version2}`);
    return response.data;
  },
};
